fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1289,24 +1294,44 @@ private boolean shouldClearExistingIndexEntries(@Nonnull IndexState indexState,
     }
 
     @Nonnull
-    private CompletableFuture<Void> doBuildIndexAsync(boolean markReadable) {
-        CompletableFuture<Void> buildFuture = buildEndpoints().thenCompose(tupleRange -> {
+    private CompletableFuture<Void> buildIndexFromRecordsAsync() {
+        return buildEndpoints().thenCompose(tupleRange -> {
             if (tupleRange != null) {
                 return buildRange(Key.Evaluated.fromTuple(tupleRange.getLow()), Key.Evaluated.fromTuple(tupleRange.getHigh()));
             } else {
                 return CompletableFuture.completedFuture(null);
             }
         });
+    }
 
-        if (markReadable) {
-            return buildFuture.thenCompose(vignore ->
-                getRunner().runAsync(context -> openRecordStore(context)
-                        .thenCompose(store -> store.markIndexReadable(index))
-                        .thenApply(ignore -> null))
-            );
-        } else {
-            return buildFuture;
-        }
+    @Nonnull
+    private CompletableFuture<Void> doBuildIndexAsync(boolean markReadable) {
+        boolean useIndexFromIndex = indexFromIndex != null && indexFromIndex.isActive();
+        CompletableFuture<Void> buildFuture =
+                ! useIndexFromIndex  ?
+                buildIndexFromRecordsAsync() :
+                buildIndexFromIndexAsync()
+                        .handle( (ignore, ex) ->
+                        {
+                            if (ex != null) {
+                                if ( indexFromIndex.isAllowRecordsScan() &&
+                                         // Should this exception be a distinct class?
+                                         ex.getMessage().contains(IndexFromIndexRequest.BAD_SRC_MESSAGE)) {
+                                    // fallback to a records scan
+                                    return buildIndexFromRecordsAsync();
```

Comment:
To my understanding - no. 
The index scan does not make any updates on the build ranges until it's all done. Therefore the record scan can continue from whatever valid state the built ranges was in, when this function was called. We can think about it this way - the built ranges only reflects the records scan indexing's state. I works because index scan is allowed only for idempotent indexes. 
I have a radar to change this later on, and let the built ranges be aware of which type of indexing state it is reflecting. I'm not sure if it'll be 'either' or 'both'.